MCA: Support repo/baseurl overrides for from and to versions

jwakre opened the pull request:
Problem:

In some cases (downstream mix with custom content) the `from` and `to` mix versions require different repo settings in the dnf conf file to download packages correctly. This makes it difficult for MCA to consume a single config file when downloading packages for both mix versions.

Approach:

Add arguments that override repo/baseurl settings in a temporary dnf config file for that will be used only when downloading packages for the `from` or `to` mixes.

If this approach is acceptable, I will update with tests

jwakre commented:
Maybe a better solution would be to use a single dnf config file and allow mixer to enable and disable repos on the command line. If there is a way to disable/enable repos separately for the from/to mixes, this would also work. I am open to other suggestions.

rchiossi commented on builder/repo_control.go:
```diff
@@ -224,6 +224,34 @@ func (b *Builder) SetExcludesRepo(reponame, pkgs string) error {
+	}
+	origConf, err := ioutil.ReadFile(b.Config.Builder.DNFConf)
+	if err != nil {
+		return tmpConf, err
```
i don't like the fact that the caller is responsible for closing tmpConf if the function fails here. This can lead to potential leak if this function is ever reused by someone that is not aware of this requirement.

jwakre replied:
I'll move this to another function.

rchiossi commented on builder/repo_control.go:
```diff
@@ -224,6 +224,34 @@ func (b *Builder) SetExcludesRepo(reponame, pkgs string) error {
+	if err != nil {
+		return tmpConf, err
+	}
+	err = ioutil.WriteFile(tmpConf.Name(), origConf, 0644)
```
You already have a file descriptor open for writing, you don't need to open it again here.

rchiossi commented on mixer/cmd/build.go:
```diff
@@ -490,13 +492,62 @@ var buildValidateCmd = &cobra.Command{
+		}
+		setWorkers(toBuilder)
+
+		if buildFlags.fromRepoURLs != nil {
```
the code below should be on a function to avoid duplication.

jwakre commented:
> Not in the scope of this PR, but to you have an idea of which parts of builder are needed for MCA? I want to measure the effort to isolating that as an initial work on refactoring `builder`.

Variables:
* b.Config.Builder.DNFConf - This can be different for from/to mix versions
* b.Config.Builder.ServerStateDir
* b.NumBundleWorkers
* b.State.Mix.Format

Functions:
* b.getLocalUpstreamVersion()
* b.isPlus10Version()
* b.checkFormatsMatch()
* b.getFormatForVersion()
* b.getLatestForFormat()

jwakre commented:
> I don't like the fact that we need two build objects here. It is just a good indication that `builder` has accumulated too many functionality and would be better split into dedicated and independent components.

I agree that 2 build objects is not ideal. Instead, I could pass the DNFConf path override directly and use a single build object. Separated components would also be nice.

jwakre commented:
@rchiossi I refactored the PR so that it no longer relies on 2 builder objects. The other feedback has been addressed too.

rchiossi commented on builder/bundles.go:
```diff
@@ -471,7 +472,7 @@ func downloadRpms(packagerCmd, rpmList []string, baseDir string, maxRetries int)
 
 	for attempts := 0; attempts <= maxRetries; attempts++ {
-		out, downloadErr := helpers.RunCommandOutputEnv(args[0], args[1:], []string{"LC_ALL=en_US.UTF-8"})
+		out, downloadErr = helpers.RunCommandOutputEnv(args[0], args[1:], []string{"LC_ALL=en_US.UTF-8"})
```
nice catch!

rchiossi commented on builder/build_validate.go:
```diff
@@ -255,14 +257,34 @@ func (b *Builder) mcaPkgInfo(mInfo map[string]*swupd.Manifest, version, download
+			return nil, err
+		}
+		defer func() {
+			_ = os.Remove(tmpConf.Name())
```
You should also close the file descriptor before removing the file.

jwakre replied:
Is it not safe to rely on the garbage collector for this?

rchiossi replied:
the GC runs concurrent to the main program, so we cannot ensure that the file descriptor will be closed before it is reused in some other function.

rchiossi commented on mixer/cmd/build.go:
```diff
@@ -788,3 +802,23 @@ func checkFormatBump(b *builder.Builder) error {
+
+// parseRepoURLOverrides parses the --from-repo-url and --to-repo-url arguments
+// to create a map of repo keys with URL values.
+func parseRepoURLOverrides(repoURLs []string) (map[string]string, error) {
```
Isn't there are facility provided by Cobra to parse a list of args like this?

jwakre replied:
You are right, the `StringToString` Pflags type can store flag arguments into a map[string]string. The flag format will change to the following:

`--from-repo-url <repo>=<url>`

Our current version of Pflags does not support this feature, so we need to bump the version.

jwakre replied:
Upgrading Cobra from 0.0.1 to 0.0.5 enables the required flags.

rchiossi commented:
One more thing: Remember to update the docs as well on those two new flags.

jwakre commented:
I updated the PR with the following changes:

* Changed the underlying Pflag type for the `--from-repo-url/--to-repo-url` args from `StringArrayVar` with `StringToString`. This allows Cobra to populate a map[string]string type. As a result, I no longer need to parse input for these flags.
* Closed the tmp config file descriptor before erasing the file.
* Added documentation.

This PR is currently blocked because a newer version of Pflag is required to use the `StringToString` flag type.
